src/common/netif.cpp

```diff
@@ -84,50 +87,78 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
 
     // Receive response.
     char response[4096];
-    int64_t recv_result;
-    do {
-        recv_result = sock->Recv(response, sizeof(response), 0);
-    } while (recv_result < 0 && (errno == EINTR || errno == EAGAIN));
-    if (recv_result < 0) {
-        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "recv() from netlink socket: %s\n", NetworkErrorString(errno));
-        return std::nullopt;
-    }
-
-    for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, recv_result); hdr = NLMSG_NEXT(hdr, recv_result)) {
-        rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
-        int remaining_len = RTM_PAYLOAD(hdr);
-
-        if (hdr->nlmsg_type != RTM_NEWROUTE) {
-            continue; // Skip non-route messages
+    ssize_t total_bytes_read{0};
+    bool done{false};
+    bool multi_part{false};
+    while (!done) {
+        int64_t recv_result;
+        do {
+            recv_result = sock->Recv(response, sizeof(response), 0);
+        } while (recv_result < 0 && (errno == EINTR || errno == EAGAIN));
+        if (recv_result < 0) {
+            LogPrintLevel(BCLog::NET, BCLog::Level::Error, "recv() from netlink socket: %s\n", NetworkErrorString(errno));
+            return std::nullopt;
         }
 
-        // Only consider default routes (destination prefix length of 0).
-        if (r->rtm_dst_len != 0) {
-            continue;
+        total_bytes_read += recv_result;
+        if (total_bytes_read > NETLINK_MAX_RESPONSE_SIZE) {
+            LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "Netlink response exceeded size limit (%zu bytes, family=%d)\n", NETLINK_MAX_RESPONSE_SIZE, family);
+            return std::nullopt;
         }
 
-        // Iterate over the attributes.
-        rtattr *rta_gateway = nullptr;
-        int scope_id = 0;
-        for (rtattr* attr = RTM_RTA(r); RTA_OK(attr, remaining_len); attr = RTA_NEXT(attr, remaining_len)) {
-            if (attr->rta_type == RTA_GATEWAY) {
-                rta_gateway = attr;
-            } else if (attr->rta_type == RTA_OIF && sizeof(int) == RTA_PAYLOAD(attr)) {
-                std::memcpy(&scope_id, RTA_DATA(attr), sizeof(scope_id));
+        bool processed_one{false};
+        for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, recv_result); hdr = NLMSG_NEXT(hdr, recv_result)) {
+            rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
+            int remaining_len = RTM_PAYLOAD(hdr);
+
+            processed_one = true;
+            if (hdr->nlmsg_flags & NLM_F_MULTI) {
+                multi_part = true;
             }
```

Comment:
nit:

`processed_one`, `multi_part` and the check at the bottom of the outer for loop could be dropped, with the following check:

```suggestion
            if (!(hdr->nlmsg_flags & NLM_F_MULTI)) {
                done = true;
            }
```

Reply:
A slightly different approach, which avoids confusing `NLMSG_DONE` with `done`:

```diff
diff --git a/src/common/netif.cpp b/src/common/netif.cpp
index 77246980c3..3f3b8fdcea 100644
--- a/src/common/netif.cpp
+++ b/src/common/netif.cpp
@@ -117,9 +117,10 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
     // Receive response.
     char response[4096];
     ssize_t total_bytes_read{0};
-    bool done{false};
-    bool multi_part{false};
-    while (!done) {
+    // If we processed at least one message and multi flag not set, or if
+    // we received no valid messages, then we're done.
+    bool maybe_more{true};
+    while (maybe_more) {
         int64_t recv_result;
         do {
             recv_result = sock->Recv(response, sizeof(response), 0);
@@ -135,18 +136,16 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
             return std::nullopt;
         }
 
-        bool processed_one{false};
         for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, recv_result); hdr = NLMSG_NEXT(hdr, recv_result)) {
             rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
             int remaining_len = RTM_PAYLOAD(hdr);
 
-            processed_one = true;
-            if (hdr->nlmsg_flags & NLM_F_MULTI) {
-                multi_part = true;
-            }
+            if (!(hdr->nlmsg_flags & NLM_F_MULTI)) {
+                maybe_more = false;
+            };
 
             if (hdr->nlmsg_type == NLMSG_DONE) {
-                done = true;
+                maybe_more = false;
                 break;
             }
 
@@ -183,12 +182,6 @@ std::optional<CNetAddr> QueryDefaultGatewayImpl(sa_family_t family)
                 }
             }
         }
-
-        // If we processed at least one message and multi flag not set, or if
-        // we received no valid messages, then we're done.
-        if ((processed_one && !multi_part) || !processed_one) {
-            done = true;
-        }
     }
 
     return std::nullopt;
```